### PR TITLE
Solve 2 issues

### PR DESCRIPTION
**Pull request by FeitianSmartcardReader**
1. Latest code has failure when generate RSA key, fix it in sm.c file
2. Solve issue->
https://github.com/OpenSC/OpenSC/issues/1073#issuecomment-309582992, add
ECC support

These code already passed test
https://github.com/OpenSC/OpenSC/wiki/Smart-Card-Testing

### CODE REVIEW

**Comment by Jakuje** [comment]
@FeitianSmartcardReader Any possibility to resolve the issue "Failed to erase card: Security status not satisfied" described in the [wiki](https://github.com/OpenSC/OpenSC/wiki/Feitian-ePass2003), without running binary blobs from internet? I have ePass token, but I am unable to us that.

**Comment by FeitianSmartcardReader** [comment]
@Jakuje The issue need using our tool can download from below and to re-format token, if you cannot using this tool, please show me your error or log, I will help check and solve it, thanks http://download.ftsafe.com/files/ePass/Fix_Tool.tar.gz

File: src/libopensc/sm.c [hunk 1]
```diff
@@ -158,7 +158,7 @@ sc_sm_single_transmit(struct sc_card *card, struct sc_apdu *apdu)
 
 	/* send APDU flagged as NO_SM */
 	sm_apdu->flags |= SC_APDU_FLAGS_NO_SM;
-	rv = sc_transmit_apdu(card, sm_apdu);
+	rv = card->reader->ops->transmit(card->reader, sm_apdu);
```
Comment (frankmorgner) [inline src/libopensc/sm.c]: This change will most likely lead to problems with other cards.

See discussion in https://github.com/OpenSC/OpenSC/issues/970

File: src/pkcs15init/pkcs15-lib.c [hunk 2]
```diff
@@ -1238,7 +1238,7 @@ sc_pkcs15init_init_prkdf(struct sc_pkcs15_card *p15card, struct sc_profile *prof
 		key_info->params.data = &keyargs->key.u.ec.params;
 		key_info->params.free_params = sc_pkcs15init_empty_callback;
 		key_info->field_length = ecparams->field_length;
-		key_info->modulus_length = 0;
+//		key_info->modulus_length = 0;
```
Comment (frankmorgner) [inline src/pkcs15init/pkcs15-lib.c]: This change will most likely break other cards

Comment (FeitianSmartcardReader) [inline src/pkcs15init/pkcs15-lib.c]: I will talk internal and make the change, thanks

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner Do modify, check changes https://github.com/OpenSC/OpenSC/pull/1085/commits/8021c472762d07b903f26e15bae755d42deeda2e

**Comment by dengert** [comment]
To avoid a strcmp, for every card, The test should be 
`switch(card->type)  {`
    `case:   SC_CARD_TYPE_ENTERSAFE_FTCOS_EPASS2003:`
        `key_info->modulus_length = 0;;`
`}`

File: src/pkcs15init/pkcs15-lib.c [hunk 3]
```diff
@@ -1238,7 +1238,13 @@ sc_pkcs15init_init_prkdf(struct sc_pkcs15_card *p15card, struct sc_profile *prof
 		key_info->params.data = &keyargs->key.u.ec.params;
 		key_info->params.free_params = sc_pkcs15init_empty_callback;
 		key_info->field_length = ecparams->field_length;
-		key_info->modulus_length = 0;
```
Comment (frankmorgner) [inline src/pkcs15init/pkcs15-lib.c]: According to ISO 7816-15 (successor of PKCS#15), an `PublicECKeyAttributes` doesn't have a `modulusLength` field (while a `
PublicRSAKeyAttributes` *does* have one):
image
image

What's the reason for this change? As far as I can see, you're not using the `modulus_length` somewhere else.

File: src/libopensc/sm.c [hunk 4]
```diff
@@ -158,7 +158,16 @@ sc_sm_single_transmit(struct sc_card *card, struct sc_apdu *apdu)
 
 	/* send APDU flagged as NO_SM */
 	sm_apdu->flags |= SC_APDU_FLAGS_NO_SM;
-	rv = sc_transmit_apdu(card, sm_apdu);
+	/* if token is epass2003, using below code*/
+	switch(card->type){
+		case SC_CARD_TYPE_ENTERSAFE_FTCOS_EPASS2003:
```
Comment (frankmorgner) [inline src/libopensc/sm.c]: I'd like to avoid card specific code on this layer. What's the reason for this change?

File: src/pkcs15init/pkcs15-lib.c (hunk 3 above)
Comment (FeitianSmartcardReader) [inline src/pkcs15init/pkcs15-lib.c]: Right, no need do modify code here, we will do it in pkcs15-epass2003.c file, will submit another pull request

File: src/libopensc/sm.c (hunk 4 above)
Comment (FeitianSmartcardReader) [inline src/libopensc/sm.c]: We double check the code, no need do modify in this layer, we will modify the file apdu.c.
The reason is our epass2003 series support plain APDU and cipher APDU, while in cipher APDU, after receive 6C command, then we will need get data back and re-package APDU with cipher mode, at this time, the set_le API seems only modify the LE, and nothing change of APDU, so we will modify apdu.c to solve it.

**Comment by sandys** [comment]
@FeitianSmartcardReader im unable to download your fix-tool from http://download.ftsafe.com/files/ePass/Fix_Tool.tar.gz. Can you post the link again please

**Comment by FeitianSmartcardReader** [comment]
check below:
[Fix_Tool.tar.gz](https://github.com/OpenSC/OpenSC/files/1192802/Fix_Tool.tar.gz)
